File: python/src/relay/workers/feed_consumer.py
"""The Kafka feed-fanout consumer group.

Delivery is at-least-once and the projector is idempotent (G-KAFKA consumer);
the offset is committed only after ``apply`` succeeds, so a processing failure
is retried (never silently skipped). Runs on a daemon thread in the harness.
"""

from __future__ import annotations

import logging
import threading

from confluent_kafka import Consumer, KafkaError

from relay import seams
from relay.infra.codecs import deserialize_message_posted

log = logging.getLogger("relay.feed_consumer")

# librdkafka's background threads emit FAIL-level connection events when the
# broker is paused (the broker-down scenarios); route them off raw stderr so
# they cannot race teardown and break pristine test output.
_rdkafka_log = logging.getLogger("relay.rdkafka")
_rdkafka_log.setLevel(logging.CRITICAL)


class FeedConsumer:
    def __init__(
        self, brokers: str, topic: str, group: str, projector: seams.FeedProjector
    ) -> None:
        self._projector = projector
        self._topic = topic
        self._stop = threading.Event()
        self._thread: threading.Thread | None = None
        self._consumer = Consumer(
            {
                "bootstrap.servers": brokers,
                "group.id": group,
                "auto.offset.reset": "earliest",
                "enable.auto.commit": False,
                "socket.timeout.ms": 1000,
                "log.connection.close": False,
            },
            logger=_rdkafka_log,
        )

    def start(self) -> None:
        self._consumer.subscribe([self._topic])
        self._thread = threading.Thread(target=self._run, name="feed-consumer", daemon=True)
        self._thread.start()

    def stop(self) -> None:
        self._stop.set()
        if self._thread is not None:
            self._thread.join(timeout=5)
        self._consumer.close()

    def _run(self) -> None:
        while not self._stop.is_set():
            message = self._consumer.poll(0.2)
            if message is None:
                continue
            if message.error() is not None:
                if message.error().code() == KafkaError._PARTITION_EOF:
                    continue
                # Broker unavailable or transient fetch error — librdkafka self-heals.
                continue
            try:
                event = deserialize_message_posted(message.value())
                self._projector.apply(event)
            except Exception:
                # Processing failed before commit: do not advance the offset; the record is
                # re-polled and retried. The projector's idempotency makes the retry safe.
                log.debug("feed fanout failed; will retry", exc_info=True)
                continue
            self._consumer.commit(message=message, asynchronous=False)

File: go/src/relay/workers/feedconsumer.go
// Package workers holds the background broker consumers: the Kafka feed-fanout consumer and
// the RabbitMQ notification worker. Both take a seam (FeedProjector / NotificationRecorder)
// so a consumer-side naive variant can be injected through the same constructor seam as the
// rest. Offsets/acks are applied only AFTER the seam's effect is persisted, so the harness's
// await-idle assertion ("group lag == 0" / "queue settled") implies the effects are durable.
package workers

import (
	"context"
	"log/slog"

	"github.com/drim-dev/verifying-agent-code/go/src/relay/infra"
	"github.com/drim-dev/verifying-agent-code/go/src/relay/seams"
	"github.com/twmb/franz-go/pkg/kgo"
)

// FeedConsumer runs the feed-fanout consumer group. Delivery is at-least-once and the
// projector is idempotent (G-KAFKA consumer); the offset is committed only after Apply
// succeeds, so a processing failure is retried (never silently skipped).
type FeedConsumer struct {
	client    *kgo.Client
	projector seams.FeedProjector
	log       *slog.Logger
}

func NewFeedConsumer(client *kgo.Client, projector seams.FeedProjector, log *slog.Logger) *FeedConsumer {
	return &FeedConsumer{client: client, projector: projector, log: log}
}

// Run consumes until ctx is cancelled. Manual commits: the offset advances only after the
// projector persisted the event's effects.
func (c *FeedConsumer) Run(ctx context.Context) {
	for {
		if ctx.Err() != nil {
			return
		}
		fetches := c.client.PollFetches(ctx)
		if ctx.Err() != nil {
			return
		}
		if errs := fetches.Errors(); len(errs) > 0 {
			// Broker unavailable or transient fetch error — franz-go self-heals; keep polling.
			continue
		}
		failed := false
		fetches.EachRecord(func(record *kgo.Record) {
			if failed {
				return
			}
			ev, err := infra.DeserializeMessagePosted(record.Value)
			if err != nil {
				c.log.Debug("skip undecodable event", "err", err)
				return
			}
			if err := c.projector.Apply(ctx, ev); err != nil {
				// Processing failed before commit: do not advance the offset; the record is
				// re-fetched and retried. The projector's idempotency makes the retry safe.
				c.log.Debug("feed fanout failed; will retry", "err", err)
				failed = true
			}
		})
		if failed {
			continue // do not commit; the uncommitted records are re-polled
		}
		if err := c.client.CommitUncommittedOffsets(ctx); err != nil {
			c.log.Debug("commit failed; will retry", "err", err)
		}
	}
}
